core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala

```diff
@@ -171,7 +171,8 @@ private[spark] class UnifiedMemoryManager private[memory] (
     if (numBytes > storagePool.memoryFree) {
       // There is not enough free memory in the storage pool, so try to borrow free memory from
       // the execution pool.
-      val memoryBorrowedFromExecution = Math.min(executionPool.memoryFree, numBytes)
+      val memoryBorrowedFromExecution = Math.min(executionPool.memoryFree,
+        numBytes - storagePool.memoryFree)
```

Comment:
This change makes sense: if we hit this branch then we've failed to acquire enough memory to store a block. In particular, we might need `M` bytes of memory but only `A` are available in storage. In this case, we should only request the `(M - A)` bytes of extra memory that we need, but the old code was incorrectly requesting all `M` which would cause us to over-evict. Therefore, I think this change is correct.